In [48]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import nltk
from nltk.stem.porter import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity




In [17]:
df=pd.read_csv('spotify_millsongdata.csv')

In [18]:
df

artist                   song  \
0              ABBA  Ahe's My Kind Of Girl   
1              ABBA       Andante, Andante   
2              ABBA         As Good As New   
3              ABBA                   Bang   
4              ABBA       Bang-A-Boomerang   
...             ...                    ...   
57645  Ziggy Marley          Good Old Days   
57646  Ziggy Marley          Hand To Mouth   
57647          Zwan           Come With Me   
57648          Zwan                 Desire   
57649          Zwan              Heartsong   

                                              link  \
0       /a/abba/ahes+my+kind+of+girl_20598417.html   
1            /a/abba/andante+andante_20002708.html   
2             /a/abba/as+good+as+new_20003033.html   
3                       /a/abba/bang_20598415.html   
4           /a/abba/bang+a+boomerang_20002668.html   
...                                            ...   
57645  /z/ziggy+marley/good+old+days_10198588.html   
57646  /z/ziggy+marley/hand+to+mouth_20531167.html   
57647           /z/zwan/come+with+me_20148981.html   
57648                 /z/zwan/desire_20148986.html   
57649              /z/zwan/heartsong_20148991.html   

                                                    text  
0      Look at her face, it's a wonderful face  \r\nA...  
1      Take it easy with me, please  \r\nTouch me gen...  
2      I'll never know why I had to go  \r\nWhy I had...  
3      Making somebody happy is a question of give an...  
4      Making somebody happy is a question of give an...  
...                                                  ...  
57645  Irie days come on play  \r\nLet the angels fly...  
57646  Power to the workers  \r\nMore power  \r\nPowe...  
57647  all you need  \r\nis something i'll believe  \...  
57648  northern star  \r\nam i frightened  \r\nwhere ...  
57649  come in  \r\nmake yourself at home  \r\ni'm a ...  

[57650 rows x 4 columns]

In [19]:
ff=df['link']

In [56]:
df.groupby('artist').count()

song  text
artist                  
'n Sync         93    93
ABBA           113   113
Ace Of Base     74    74
Adam Sandler    70    70
Adele           54    54
...            ...   ...
Zoegirl         38    38
Zornik          12    12
Zox             21    21
Zucchero        30    30
Zwan            14    14

[643 rows x 2 columns]

In [54]:
pp

NOW that we have assigned the link to another column we can drop them


In [20]:
df.drop('link', axis=1, inplace=True)


In [21]:
df.head(8)

artist                   song  \
0   ABBA  Ahe's My Kind Of Girl   
1   ABBA       Andante, Andante   
2   ABBA         As Good As New   
3   ABBA                   Bang   
4   ABBA       Bang-A-Boomerang   
5   ABBA     Burning My Bridges   
6   ABBA              Cassandra   
7   ABBA             Chiquitita   

                                                text  
0  Look at her face, it's a wonderful face  \r\nA...  
1  Take it easy with me, please  \r\nTouch me gen...  
2  I'll never know why I had to go  \r\nWhy I had...  
3  Making somebody happy is a question of give an...  
4  Making somebody happy is a question of give an...  
5  Well, you hoot and you holler and you make me ...  
6  Down in the street they're all singing and sho...  
7  Chiquitita, tell me what's wrong  \r\nYou're e...

In [22]:
df['text'][0]

"Look at her face, it's a wonderful face  \r\nAnd it means something special to me  \r\nLook at the way that she smiles when she sees me  \r\nHow lucky can one fellow be?  \r\n  \r\nShe's just my kind of girl, she makes me feel fine  \r\nWho could ever believe that she could be mine?  \r\nShe's just my kind of girl, without her I'm blue  \r\nAnd if she ever leaves me what could I do, what could I do?  \r\n  \r\nAnd when we go for a walk in the park  \r\nAnd she holds me and squeezes my hand  \r\nWe'll go on walking for hours and talking  \r\nAbout all the things that we plan  \r\n  \r\nShe's just my kind of girl, she makes me feel fine  \r\nWho could ever believe that she could be mine?  \r\nShe's just my kind of girl, without her I'm blue  \r\nAnd if she ever leaves me what could I do, what could I do?\r\n\r\n"

In [23]:
artist_list=df.groupby('artist').count()

Now we are having 643 artist list in our dataset


In [24]:
artist_list.count()

song    643
text    643
dtype: int64

In [25]:
df['text']

0        Look at her face, it's a wonderful face  \r\nA...
1        Take it easy with me, please  \r\nTouch me gen...
2        I'll never know why I had to go  \r\nWhy I had...
3        Making somebody happy is a question of give an...
4        Making somebody happy is a question of give an...
                               ...                        
57645    Irie days come on play  \r\nLet the angels fly...
57646    Power to the workers  \r\nMore power  \r\nPowe...
57647    all you need  \r\nis something i'll believe  \...
57648    northern star  \r\nam i frightened  \r\nwhere ...
57649    come in  \r\nmake yourself at home  \r\ni'm a ...
Name: text, Length: 57650, dtype: object

In [26]:

p=df['text'].str.count('\n').sum()

In [27]:
df['text'] = df['text'].apply(lambda x: str(x).lower().replace(r'^\w\s',' ').replace(r'\n',' ',))


In [32]:
df['text'][0]

"look at her face, it's a wonderful face  \r\nand it means something special to me  \r\nlook at the way that she smiles when she sees me  \r\nhow lucky can one fellow be?  \r\n  \r\nshe's just my kind of girl, she makes me feel fine  \r\nwho could ever believe that she could be mine?  \r\nshe's just my kind of girl, without her i'm blue  \r\nand if she ever leaves me what could i do, what could i do?  \r\n  \r\nand when we go for a walk in the park  \r\nand she holds me and squeezes my hand  \r\nwe'll go on walking for hours and talking  \r\nabout all the things that we plan  \r\n  \r\nshe's just my kind of girl, she makes me feel fine  \r\nwho could ever believe that she could be mine?  \r\nshe's just my kind of girl, without her i'm blue  \r\nand if she ever leaves me what could i do, what could i do?\r\n\r\n"

In [33]:
df.columns

Index(['artist', 'song', 'text'], dtype='object')

In [50]:

# import nltk
# it is a nlp library , it is used to process the text data
# from nltk.stem.porter import PorterStemmer
stemmer=PorterStemmer()

In this the similar words which are having same meaning but differ in spelling are changed into one common word


In [43]:
def token(text):
    token=nltk.word_tokenize(text)
    a=[stemmer.stem(word) for word in token]
    return " ".join(a)
    

In [47]:
# import nltk
# nltk.download('punkt')
token(df['text'][0])
token('hello world beautiful beauty')

'hello world beauti beauti'

In [51]:
df['text'].apply(lambda x:token(x))

0        look at her face , it 's a wonder face and it ...
1        take it easi with me , pleas touch me gentli l...
2        i 'll never know whi i had to go whi i had to ...
3        make somebodi happi is a question of give and ...
4        make somebodi happi is a question of give and ...
                               ...                        
57645    iri day come on play let the angel fli let the...
57646    power to the worker more power power to the wo...
57647    all you need is someth i 'll believ flashlight...
57648    northern star am i frighten where can i go to ...
57649    come in make yourself at home i 'm a bit late ...
Name: text, Length: 57650, dtype: object

In [57]:
# from sklearn.feature_extraction.text import TfidfVectorizer
# from sklearn.metrics.pairwise import cosine_similarity



TfidfVectorizer(analyzer='word', stop_words='english')

TfidfVectorizer(stop_words='english')